### ISSUE TITLE
[Bug]: shipped ref.json does not match OLMoE-1B-7B-0125-Instruct — olmoe validator reports 5/12 out of the box (engine itself is token-exact vs a fresh oracle)

### ISSUE THREAD

**Issue by lEWFkRAD** [JustVugg/colibri#133]
### Commit or version

97c756a06472b2f632453f35a541b2f0553bec00

### Environment

Windows 11 host / WSL2 Ubuntu, Intel Core Ultra 9 285K (24c, AVX2), 31 GB VM RAM, model on ext4 (NVMe), gcc 15.2.0. Oracle: transformers 5.13.1 + torch 2.13.0 (CPU, bf16).

### Reproduction steps

```sh
make -C c olmoe
# model: allenai/OLMoE-1B-7B-0125-Instruct (bf16 safetensors, 3 shards)
SNAP=/path/olmoe OMP_NUM_THREADS=16 ./c/olmoe 64 8 ref.json
```

### Expected behavior

`Matching tokens: 12/12` against the shipped `ref.json`, per the engine's token-exact acceptance criterion.

### Actual behavior and logs

```
Reference: 7785 15 187 187 510 5347 273 253 1986 2077 310 5041
C engine : 7785 15 187 187 510 3565 3448 273 6181 310 5112 15
Matching tokens: 5/12
```

**The engine is not at fault — the shipped `ref.json` appears to be generated from a different checkpoint.** I regenerated the oracle independently (transformers 5.13.1, bf16 CPU, greedy) on the exact checkpoint above:

- oracle greedy continuation: ` Paris.\n\nThe official language of France is French.` = tokens `[7785, 15, 187, 187, 510, 3565, 3448, 273, 6181, 310, 5112, 15]`
- **oracle vs C engine (int8, cap=64): 12/12 — token-exact.** Also 12/12 at cap=8 (disk-streamed, 18% cache hit), i.e. the streaming path is exact too.
- oracle vs shipped `ref.json`: 5/12.
- Teacher-forced argmax over the shipped `full_ids` on this checkpoint: 9/12, first divergence at position 10 (`' capital'` expected by ref vs `' official'` argmax) — this checkpoint's greedy path cannot produce the shipped continuation.

So the engine passes its own validation, but a fresh clone + the documented checkpoint reports 5/12 out of the box, which reads as an engine failure.

Suggested fix: regenerate `ref.json` against a pinned checkpoint and record provenance in the file (repo id, revision hash, transformers version, dtype, device) so the fixture is auditable. Happy to PR the regenerated fixture if you tell me which snapshot you consider canonical.

**Comment by lEWFkRAD** [comment]
Follow-up with stronger evidence for "the engine is fine, the fixture is stale":

**Multi-prompt battery** — 8 diverse prompts (factual / code / narrative / numeric), 24 greedy tokens each, references regenerated with a provenance-stamped transformers oracle (allenai/OLMoE-1B-7B-0125-Instruct, bf16, CPU, transformers 5.13.1, torch 2.13.0), engine at int8 cap=64:

- **7/8 prompts: 24/24 token-exact.**
- The single divergence ("In 1969, the first humans landed on", position 4) is an **exact argmax tie**: the oracle's top-2 logits are bit-identical in bf16 (`' The'` = `' Neil'` = 13.4375, gap 0.0000). The engine picked the other member of the tie — both are valid greedy outputs. So: **8/8 under tie-equivalence.**

Two practical suggestions for the regenerated fixture, whenever you pin a canonical snapshot:
1. Record provenance in `ref.json` (repo id, revision, transformers/torch versions, dtype, device).
2. Have the generator check the top-1/top-2 logit gap at every emitted position and reject prompts containing near-ties (or record the tie set) — otherwise the fixture can "fail" on engines that are exactly correct.

Happy to PR the regenerated multi-prompt fixture + a tie-gap check in the generator.

**Comment by JustVugg** [comment]
Fixed in `main` (`1f0e1b7`). You're right — the engine was never at fault. The shipped `ref.json` continued *'The capital of the United States is Washington'* (a different checkpoint), while the intended target per `convert_olmoe.py` is `allenai/OLMoE-1B-7B-0125-Instruct`, whose greedy oracle continues *'The official language of France is French'* — exactly your regenerated tokens. I updated `full_ids`/`text` to your verified oracle `[7785, 15, 187, 187, 510, 3565, 3448, 273, 6181, 310, 5112, 15]`. Since the engine is token-exact vs that oracle (your 12/12), the validator should now report **12/12** out of the box. I can't re-run it here (no OLMoE checkout), so if you get a moment: `SNAP=/path/olmoe ./c/olmoe 64 8 ref.json` should be 12/12 now — a quick confirm would close the loop. Thank you for doing the oracle work and separating the two bugs so cleanly.

**Comment by lEWFkRAD** [comment]
Confirmed on `1f0e1b7` against the same OLMoE-1B-7B-0125-Instruct checkout (WSL2 Ubuntu, Ultra 9 285K, gcc 15.2, `make olmoe ARCH=native`):

```
$ SNAP=/path/olmoe OMP_NUM_THREADS=16 ./olmoe 64 8 ref.json
Matching tokens: 12/12
```

Also verified the #134 fix from the same commit:

```
$ ./olmoe 64 16 ref.json
quant_bits must be 2..8 (got 16); OLMoE experts are int8-backed, no f32 mode
```

Validator now passes out of the box. Loop closed — thanks for the fast turnaround.

**Comment by JustVugg** [comment]
thanks for your help for improve colibrì!
